File: backend/app/api/v1/export.py
"""Data export API endpoints.

POST /api/v1/dbs/{name}/export re-executes the query server-side and
returns the result as a downloadable CSV / JSON file (streamed so large
result sets do not need to be held in memory twice).
"""

import logging

from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import StreamingResponse
from sqlmodel import Session, select

from app.config import settings
from app.database import get_session
from app.models.database import DatabaseConnection
from app.models.query import QuerySource
from app.models.schemas import ExportRequest
from app.services.export_service import (
    build_filename,
    get_media_type,
    stream_export,
)
from app.services.query_wrapper import execute_query_with_service
from app.services.sql_validator import SqlValidationError

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/dbs", tags=["export"])


@router.post("/{name}/export")
async def export_query_results(
    name: str,
    input_data: ExportRequest,
    session: Session = Depends(get_session),
) -> StreamingResponse:
    """
    Execute a SQL query and download the result as CSV or JSON.

    Args:
        name: Database connection name
        input_data: Export request (sql + format + optional limit)
        session: Database session

    Returns:
        StreamingResponse with the file content and attachment headers

    Raises:
        HTTPException 404: Connection not found
        HTTPException 400: SQL validation error
        HTTPException 500: Query execution / export failure
    """
    # Verify the connection exists
    statement = select(DatabaseConnection).where(DatabaseConnection.name == name)
    connection = session.exec(statement).first()

    if not connection:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Database connection '{name}' not found",
        )

    # The SQL validator only applies the limit when the query has no LIMIT clause,
    # so exporting a query that already limits its rows keeps that behavior.
    effective_limit = input_data.limit or settings.query_default_limit

    # Execute the query server-side, reusing the standard execution chain
    # (API -> query wrapper -> DatabaseService -> adapter).
    try:
        result = await execute_query_with_service(
            session,
            name,
            connection.db_type,
            connection.url,
            input_data.sql,
            QuerySource.MANUAL,
            limit=effective_limit,
        )
    except SqlValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        logger.exception("Export query failed for database '%s'", name)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Export query execution failed: {str(e)}",
        )

    columns = [{"name": col.name, "dataType": col.data_type} for col in result.columns]
    rows = result.rows
    export_format = input_data.format

    filename = build_filename(name, export_format)

    logger.info(
        "Exporting %d rows (%s) for database '%s'", len(rows), export_format.value, name
    )

    return StreamingResponse(
        stream_export(columns, rows, export_format),
        media_type=get_media_type(export_format),
        headers={
            "Content-Disposition": f'attachment; filename="{filename}"',
            "X-Export-Row-Count": str(len(rows)),
        },
    )

File: frontend/src/services/export.ts
/**
 * Query result export service.
 *
 * Delegates export to the backend endpoint (POST /api/v1/dbs/{name}/export)
 * so large result sets are streamed server-side instead of being built in
 * the browser. The response is a file download (Blob + <a> click).
 */

export type ExportFormat = "csv" | "json";

const API_BASE_URL =
  import.meta.env.VITE_API_BASE_URL || "http://localhost:8000";

function getFilenameFromDisposition(disposition: string | null, fallback: string): string {
  if (!disposition) return fallback;
  const match = /filename="?([^"]+)"?/.exec(disposition);
  return match?.[1] || fallback;
}

function buildFallbackFilename(databaseName: string, format: ExportFormat): string {
  const timestamp = new Date()
    .toISOString()
    .replace(/[:.]/g, "-")
    .slice(0, 19);
  return `${databaseName}_${timestamp}.${format}`;
}

/**
 * Execute a query server-side and download the result as a file.
 *
 * @param databaseName  Database connection name
 * @param sql           SQL SELECT query to export
 * @param format        "csv" | "json"
 * @param limit         Optional row limit (only applies when SQL has no LIMIT)
 */
export async function exportQuery(
  databaseName: string,
  sql: string,
  format: ExportFormat,
  limit?: number
): Promise<{ filename: string; rowCount: number }> {
  const response = await fetch(
    `${API_BASE_URL}/api/v1/dbs/${encodeURIComponent(databaseName)}/export`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ sql, format, limit }),
    }
  );

  if (!response.ok) {
    let detail = `Export failed (HTTP ${response.status})`;
    try {
      const data = await response.json();
      detail = data?.detail || detail;
    } catch {
      // Non-JSON error body - keep the generic message.
    }
    throw new Error(detail);
  }

  const blob = await response.blob();
  const fallback = buildFallbackFilename(databaseName, format);
  const filename = getFilenameFromDisposition(
    response.headers.get("Content-Disposition"),
    fallback
  );
  const rowCount = Number(response.headers.get("X-Export-Row-Count")) || 0;

  const link = document.createElement("a");
  link.href = URL.createObjectURL(blob);
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  link.remove();
  URL.revokeObjectURL(link.href);

  return { filename, rowCount };
}
